[DNM] Implements IP utilization tracking for networks

aepifanov opened the pull request:
**- What I did**

This PR introduces IP utilization tracking for Docker networks by extending the network inspection API to include IPAM state. This enhancement allows users to monitor how many IP addresses are allocated per subnet and IP range pool within a network.
It covers only IPv4.

**- How I did it**

[Design IP-utilization](https://docs.google.com/document/d/1CmmCe1T8ze8tAOyl3MKsJj_1hVlzpSGK7D4kclS9PW0/edit?tab=t.0)

1. Added IP allocation counters to track usage within:
    - The entire subnet
    - The defined IP range pool
2. Introduced new API types to represent network `State`, including nested `IPAM` state information
3. Implemented support and test coverage for the following network drivers:
    - `bridge`
    - `macvlan`
    - `ipvlan`
    - `overlay`

Example output from `docker network inspect`:
```bash
> docker network inspect my
[
    {
        "Name": "my",
....
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "192.168.10.0/24",
                    "IPRange": "192.168.10.128/31",
                    "Gateway": "192.168.10.1",
                    "AuxiliaryAddresses": {
                        "my-host": "192.168.10.128"
                    }
                }
            ]
        },
        "State": {
            "IPAM": {
                "192.168.10.0/24": {
                    "AllocatedIPsInSubnet": 4,
                    "AvailableIPsInIPRangePool": 1
                }
            }
        },
....
    }
]
```

- **AllocatedIPsInSubnet**: Number of addresses allocated in the subnet. If the value exceeds `uint64`, it is capped at the maximum `uint64` value.

- **AvailableIPsInIPRangePool**: Number of available addresses for allocating in the IP range pool.

Depends on:

[IP utilization in Swarm](https://github.com/moby/swarmkit/pull/3209)

**- How to verify it**

Run `docker network inspect <network>` and verify the new `State.IPAM` section reflects real-time allocation counts for each subnet configuration.

**- Human readable description for the release notes**

```markdown changelog
Implements IP utilization tracking for Docker networks.
Adds IP allocation counters to network inspect responses.
```

**- A picture of a cute animal (not mandatory but encouraged)**


<img width="1024" height="1536" alt="image" src="https://github.com/user-attachments/assets/b50c2cb4-525b-4230-96d3-6adb543e80ca" />

aepifanov committed:
bitmap: Add CalculatedSelected method

Adds a new method:

- CalculatedSelected(start, end uint64) (uint64, error)

    Returns the number of selected bits within the specified bit range
    (start to end, inclusive). Useful for calculating IP utilization
    within a given range.

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov committed:
api network: Extend Network inspection with State.IPAM

Extend network inspection with State object which includes IPAM state with ip utilization details.

Add a new method to the Ipam interface:

- GetAllocatedIPs(poolID string) (allocatedIPsInSubnet, allocatedIPsInPool uint64, err error)

    Returns the number of addresses allocated in both the subnet and the ip-range pool.

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov committed:
overlay: Add State.IPAM support in Network inspection

Adds support for inspecting overlay networks with a State object
that includes IPAM state and IP utilization details.

To enable this, the cnmallocator is extended with a new
UpdateNetworkState() method, used in a callback from the Swarm API
server to populate IPAM data in the network State object. This
State object is passed to the API server as a blob, avoiding
further changes to the Swarm API and preserving backward
compatibility in future releases.

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

# Conflicts:
#	go.mod

aepifanov committed:
addspace: Add GetAllocatedIPs() method

Adds a new method:

- GetAllocatedIPs(nw, ipr netip.Prefix) (allocatedIPsInSubnet, allocatedIPsInPool uint64, err error)

    Returns the number of addresses allocated in both the subnet and the ip-range pool.

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov committed:
ipvlan: Add test for ip utilization counters

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov committed:
bridge: Add test for ip utilization counters

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov committed:
macvlan: Add test for ip utilization counters

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov committed:
overlay: Add test for ip utilization counters

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

corhere reviewed (commented):
How come you need both the `PoolData.allocatedIPsInPool` counters and `(*addrset.AddrSet).Selected()`? Aren't all reserved IPs added to an AddrSet? I would have expected `.Selected()` to be sufficient.

corhere commented:
Okay, I think I understand. I am not entirely thrilled with there being two sources of truth for the IP address allocations, but you may have had a good reason to take this approach. Which alternative solutions did you consider and why did the approach of maintaining a running counter win out?

aepifanov commented:
@corhere I've done the following changes from the last your review:
1. removed ipv6 support
2. reversed the counter for ip-range pool from used to available
3. extened the integratio tests with using aux-addresses
4. changed the State.IPAM format to:
```
        "State": {
            "IPAM": {
                "192.168.10.0/24": {
                    "AllocatedIPsInSubnet": 5,
                    "AvailableIPsInIPRangePool": 1
                }
        },
```

corhere commented on daemon/libnetwork/network.go:
```diff
@@ -2193,3 +2194,59 @@ func (n *Network) deleteLoadBalancerSandbox() error {
+			if !cidr.Addr().Is4() {
+				return 0, fmt.Errorf("IPAM state only supports IPv4 CIDRs, got %s", cidr)
+			}
+			return uint64(1) << (32 - cidr.Bits()), nil
```
We can do better than hardcode a constant:
```suggestion
			return uint64(1) << (cidr.Addr().BitLen() - cidr.Bits()), nil
```

aepifanov replied:
Done

corhere commented on daemon/libnetwork/ipams/defaultipam/address_space_test.go:
```diff
@@ -498,3 +498,119 @@ func TestPoolAllocateAndRelease(t *testing.T) {
+			as, err := newAddrSpace(tc.predefined)
+			assert.NilError(t, err)
+
+			err = as.allocateSubnet(tc.subnet, tc.ipRange)
```
Where are you testing the cases with multiple ip-range allocations under a single subnet? Be sure to cover both the disjoint ranges (e.g. `--subnet 192.168.0.0/24 --ip-range 192.168.0.0/25`, `--subnet 192.168.0.0/24 --ip-range 192.168.0.128/25`) [and overlapping ranges](https://github.com/moby/moby/issues/46756) (e.g. `--subnet 192.168.0.0/24 --ip-range 192.168.0.0/25`, `--subnet 192.168.0.0/24 --ip-range 192.168.0.64/26`) cases

corhere reviewed (changes requested):
It does not look like the tests cover all the corner cases.

aepifanov committed:
addrset: Add methods to count selected addresses

Adds the following new methods:

- Selected() uint64

    Returns the number of selected addresses in the set.

- CalculateSelectedInRange(netip.Prefix) (uint64, error)

    Returns the number of selected addresses in the ip-range.

- CalculateSelected() (uint64, error)

    Returns the number of selected addresses in the set.

Signed-off-by: Andrey Epifanov <aepifanov@mirantis.com>

aepifanov commented on daemon/libnetwork/internal/addrset/addrset.go:
```diff
@@ -174,6 +175,35 @@ func (as *AddrSet) getBitmap(addr netip.Addr) (*bitmap.Bitmap, netip.Prefix, err
+	for _, bm := range as.bitmaps {
+		if bms > 0 {
+			bms--
+		}
```
Outdated code. Removed.

aepifanov commented on daemon/libnetwork/ipams/defaultipam/structures.go:
```diff
@@ -21,6 +22,9 @@ type PoolData struct {
 	children map[netip.Prefix]struct{}
 
+	// The number of addresses allocated in the pool.
+	allocatedIPsInPool uint64
```
Nice catch!
I've re-implemented `usedInIPRangePool` handling method to the direct calculation of the selected addresses in a bitmap for a specified range, which allows to get the actual IPAM state for any range.

aepifanov commented on daemon/libnetwork/ipams/defaultipam/allocator.go:
```diff
@@ -328,3 +328,25 @@ func getAddress(base netip.Prefix, addrSet *addrset.AddrSet, prefAddress netip.A
+//     and should no longer be modified.
+func (a *Allocator) GetAllocatedIPs(poolID string) (allocatedIPsInSubnet, allocatedIPsInPool uint64, err error) {
+	log.G(context.TODO()).Debugf("GetAllocatedIPs(%s)", poolID)
+	k, err := PoolIDFromString(poolID)
```
In the current implementation the `range` is taken from the `PoolID` which contains it.

corhere commented on api/types/network/network.go:
```diff
@@ -80,6 +99,7 @@ type Inspect struct {
 	EnableIPv4 bool                        // EnableIPv4 represents whether IPv4 is enabled
 	EnableIPv6 bool                        // EnableIPv6 represents whether IPv6 is enabled
 	IPAM       IPAM                        // IPAM is the network's IP Address Management
+	State      *NetworkState               `json:",omitempty"` // State represents the state of the network
```
Please update the Swagger spec and API changelog with this addition
